### ISSUE TITLE
ApiAuthException 'You must be signed-in with Cognito Userpools to be able to use getTokens'

### ISSUE THREAD

**Issue by two4zero** [aws-amplify/amplify-flutter#1220]
**Describe the bug**

There was an amplify backend in development. It worked normally until 13 noon. (2021-12-17 13:00. gmt+9)
After that, ApiAuthException suddenly occurred.
Even if I create a new flutter project and connect it, an error occurs.

```
Syncing files to device Android SDK built for x86...
I/amplify:flutter:api( 3471): Added API plugin
I/amplify:flutter:auth_cognito( 3471): Added Auth plugin
D/AWSMobileClient( 3471): Using the SignInProviderConfig from `awsconfiguration.json`.
D/AWSMobileClient( 3471): Inspecting user state details
W/AWSMobileClient( 3471): signalTokensNotAvailable
W/InternalCallback( 3471): Library attempted to call user callback twice, expected only once
I/flutter ( 3471): AmplifyException(message: Failed to retrieve auth token from Cognito provider.,
 recoverySuggestion: Check the application logs for details.,
 underlyingException: ApiAuthException{message=You must be signed-in with Cognito Userpools to be able to use getTokens,
   cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.})
```

**To Reproduce**
1. create new flutter project in android studio

 ```
Future<void> _configureAmplify() async {
    await Amplify.addPlugins([
      AmplifyAuthCognito(), // add first
      AmplifyAPI(), // add second
    ]);

    try {
      await Amplify.configure(amplifyconfig);
    } catch (e) {
      print(e);
    }
  }
```

2. just add 'AmplifyAuthCognito()' > no error
3. next. add 'AmplifyAPI()'

**Expected behavior**

**Screenshots**

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[v] Android
[] iOS

<details>

  <summary>Output of <code>flutter doctor -v</code></summary>
  
  ```
>flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 2.8.1, on Microsoft Windows [Version 10.0.22000.376], locale ko-KR)
[√] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2022 17.0.4)
[√] Android Studio (version 2020.3)
[√] VS Code (version 1.63.1)
[√] Connected device (4 available)

• No issues found!
```

</details>

<details>

  <summary>Dependencies (<code>pubspec.lock</code>)</summary>
  
  ```
# Generated by pub
# See https://dart.dev/tools/pub/glossary#lockfile
packages:
  amplify_analytics_plugin_interface:
    dependency: transitive
    description:
      name: amplify_analytics_plugin_interface
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  amplify_api:
    dependency: "direct main"
    description:
      name: amplify_api
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  amplify_api_plugin_interface:
    dependency: transitive
    description:
      name: amplify_api_plugin_interface
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  amplify_auth_cognito:
    dependency: "direct main"
    description:
      name: amplify_auth_cognito
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  amplify_auth_plugin_interface:
    dependency: transitive
    description:
      name: amplify_auth_plugin_interface
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  amplify_core:
    dependency: transitive
    description:
      name: amplify_core
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  amplify_datastore:
    dependency: "direct main"
    description:
      name: amplify_datastore
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  amplify_datastore_plugin_interface:
    dependency: transitive
    description:
      name: amplify_datastore_plugin_interface
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  amplify_flutter:
    dependency: "direct main"
    description:
      name: amplify_flutter
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  amplify_storage_plugin_interface:
    dependency: transitive
    description:
      name: amplify_storage_plugin_interface
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.2.10"
  async:
    dependency: transitive
    description:
      name: async
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.8.2"
  boolean_selector:
    dependency: transitive
    description:
      name: boolean_selector
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.1.0"
  characters:
    dependency: transitive
    description:
      name: characters
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.2.0"
  charcode:
    dependency: transitive
    description:
      name: charcode
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.3.1"
  clock:
    dependency: transitive
    description:
      name: clock
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.1.0"
  collection:
    dependency: transitive
    description:
      name: collection
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.15.0"
  crypto:
    dependency: transitive
    description:
      name: crypto
      url: "https://pub.dartlang.org"
    source: hosted
    version: "3.0.1"
  cupertino_icons:
    dependency: "direct main"
    description:
      name: cupertino_icons
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.4"
  date_time_format:
    dependency: transitive
    description:
      name: date_time_format
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.0.1"
  fake_async:
    dependency: transitive
    description:
      name: fake_async
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.2.0"
  flutter:
    dependency: "direct main"
    description: flutter
    source: sdk
    version: "0.0.0"
  flutter_lints:
    dependency: "direct dev"
    description:
      name: flutter_lints
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.4"
  flutter_test:
    dependency: "direct dev"
    description: flutter
    source: sdk
    version: "0.0.0"
  lints:
    dependency: transitive
    description:
      name: lints
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.1"
  matcher:
    dependency: transitive
    description:
      name: matcher
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.12.11"
  meta:
    dependency: transitive
    description:
      name: meta
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.7.0"
  path:
    dependency: transitive
    description:
      name: path
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.8.0"
  plugin_platform_interface:
    dependency: transitive
    description:
      name: plugin_platform_interface
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.0.2"
  sky_engine:
    dependency: transitive
    description: flutter
    source: sdk
    version: "0.0.99"
  source_span:
    dependency: transitive
    description:
      name: source_span
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.8.1"
  stack_trace:
    dependency: transitive
    description:
      name: stack_trace
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.10.0"
  stream_channel:
    dependency: transitive
    description:
      name: stream_channel
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.1.0"
  string_scanner:
    dependency: transitive
    description:
      name: string_scanner
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.1.0"
  term_glyph:
    dependency: transitive
    description:
      name: term_glyph
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.2.0"
  test_api:
    dependency: transitive
    description:
      name: test_api
      url: "https://pub.dartlang.org"
    source: hosted
    version: "0.4.3"
  typed_data:
    dependency: transitive
    description:
      name: typed_data
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.3.0"
  uuid:
    dependency: transitive
    description:
      name: uuid
      url: "https://pub.dartlang.org"
    source: hosted
    version: "3.0.5"
  vector_math:
    dependency: transitive
    description:
      name: vector_math
      url: "https://pub.dartlang.org"
    source: hosted
    version: "2.1.1"
sdks:
  dart: ">=2.15.1 <3.0.0"
  flutter: ">=1.20.0"
  ```

</details>

**Smartphone (please complete the following information):**

**Additional context**

**Comment by HuiSF** [comment]
Hi @two4zero thanks for reporting this issue, could you also share the generated `amplifyconfiguration.dart`? (Please be sure mask sensitive fields)

**Comment by two4zero** [comment]
Hi @HuiSF 

thanks~

```
const amplifyconfig = ''' {
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "xxx": {
                    "endpointType": "GraphQL",
                    "endpoint": "https://xxx.appsync-api.ap-northeast-2.amazonaws.com/graphql",
                    "region": "ap-northeast-2",
                    "authorizationType": "API_KEY",
                    "apiKey": "xxx"
                },
                "AdminQueries": {
                    "endpointType": "REST",
                    "endpoint": "https://xxx.execute-api.ap-northeast-2.amazonaws.com/staging",
                    "region": "ap-northeast-2",
                    "authorizationType": "AMAZON_COGNITO_USER_POOLS"
                }
            }
        }
    },
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "UserAgent": "aws-amplify-cli/0.1.0",
                "Version": "0.1.0",
                "IdentityManager": {
                    "Default": {}
                },
                "AppSync": {
                    "Default": {
                        "ApiUrl": "https://xxx.appsync-api.ap-northeast-2.amazonaws.com/graphql",
                        "Region": "ap-northeast-2",
                        "AuthMode": "API_KEY",
                        "ApiKey": "xxx",
                        "ClientDatabasePrefix": "xxx_API_KEY"
                    },
                    "xxx_AWS_IAM": {
                        "ApiUrl": "https://xxx.appsync-api.ap-northeast-2.amazonaws.com/graphql",
                        "Region": "ap-northeast-2",
                        "AuthMode": "AWS_IAM",
                        "ClientDatabasePrefix": "xxx_AWS_IAM"
                    },
                    "xxx_AMAZON_COGNITO_USER_POOLS": {
                        "ApiUrl": "https://xxx.appsync-api.ap-northeast-2.amazonaws.com/graphql",
                        "Region": "ap-northeast-2",
                        "AuthMode": "AMAZON_COGNITO_USER_POOLS",
                        "ClientDatabasePrefix": "xxx_AMAZON_COGNITO_USER_POOLS"
                    }
                },
                "CredentialsProvider": {
                    "CognitoIdentity": {
                        "Default": {
                            "PoolId": "ap-northeast-2:xxx",
                            "Region": "ap-northeast-2"
                        }
                    }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "ap-northeast-2_xxx",
                        "AppClientId": "xxx",
                        "Region": "ap-northeast-2"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "USER_SRP_AUTH",
                        "socialProviders": [],
                        "usernameAttributes": [
                            "EMAIL"
                        ],
                        "signupAttributes": [
                            "EMAIL",
                            "NAME"
                        ],
                        "passwordProtectionSettings": {
                            "passwordPolicyMinLength": 8,
                            "passwordPolicyCharacters": []
                        },
                        "mfaConfiguration": "OFF",
                        "mfaTypes": [
                            "SMS"
                        ],
                        "verificationMechanisms": [
                            "EMAIL"
                        ]
                    }
                }
            }
        }
    }
}''';

```

**Comment by two4zero** [comment]
Hi @HuiSF

### i think it's a 'Admin queries API' issue.
and, I must use 'Admin queries'.

### I did some tests.

1. Create a new amplify app in aws console.
2. new flutter app in android studio

3. amplify add auth - with 'Default configuration'
4. amplify push > **ok. 'Amplify.configure()' no errors**

5. amplify update auth - with 'Create or update Cognito user pool groups'
add 'Admins' group
6. amplify push > **ok. 'Amplify.configure()' no errors**

7. **amplify update auth - with 'Create or update Admin queries API'**
8. amplify push > error. 'Amplify.configure()' error occured
```
Syncing files to device Android SDK built for x86...
I/amplify:flutter:api( 6710): Added API plugin
I/amplify:flutter:auth_cognito( 6710): Added Auth plugin
D/AWSMobileClient( 6710): Using the SignInProviderConfig from `awsconfiguration.json`.
D/AWSMobileClient( 6710): Inspecting user state details
W/AWSMobileClient( 6710): signalTokensNotAvailable
W/InternalCallback( 6710): Library attempted to call user callback twice, expected only once
I/flutter ( 6710): AmplifyException(message: Failed to retrieve auth token from Cognito provider.,
 recoverySuggestion: Check the application logs for details.,
 underlyingException: ApiAuthException{message=You must be signed-in with Cognito Userpools to be able to use getTokens,
 cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.})
Application finished.
```

**Comment by two4zero** [comment]
OMG...

12.20. 17:00 gmt+9

no changes in front/backend
it's working...

thanks. ㅠ

**Comment by two4zero** [comment]
@Sunsiha I think... I haven't done anything.
Just waited 3 days.
For 3 days, I deleted 'awsAPIPlugin.AdminQueries' field from 'amplifyconfiguration.dart' and developed it for testing.

**Comment by HuiSF** [comment]
Hi @two4zero sorry I've never used Admin Queries API myself 😅 I'll give it try. 

To help us produce this issue, could you share your code example where you are using the Admin Queries?

Also could you summarize the auth flow of your App, e.g.

When do you send Admin Queries? After user signing in? Out of signed user session etc.

**Comment by two4zero** [comment]
@HuiSF 
To Reproduce >> https://github.com/aws-amplify/amplify-flutter/issues/1220#issuecomment-997560432

sorry. No shared code examples.
I just made it according to the manual.
And, an error occurred in **'Amplify.configure()'**. It is not an error when sending 'admin query' or 'sign in'.
- https://docs.amplify.aws/start/getting-started/add-api/q/integration/flutter/#add-api-and-auth-plugins
- https://docs.amplify.aws/cli/auth/admin/#enable-admin-queries

But now it works normally.

cc @Sunsiha

**Comment by HuiSF** [comment]
I can reproduce this error following @two4zero provided steps.
Will keep looking into the issue.

**Comment by HuiSF** [comment]
Hi @two4zero to clarify regarding your point -

>And, an error occurred in 'Amplify.configure()'. It is not an error when sending 'admin query' or 'sign in'.

'admin query' or 'sign in' were working even when the exception was thrown?

**Comment by HuiSF** [comment]
This exception seems not directly related to using Admin Queries API, it happens as long as when the API is REST. 

The exception was thrown at this point:

https://github.com/aws-amplify/amplify-android/blob/17fed06812c7a9ca65c6d47e7c5dbaed18846ed0/aws-api/src/main/java/com/amplifyframework/api/aws/AWSApiPlugin.java#L157-L159

It invokes a code block where it seems not handling the situation for the case that there is no user signed in

https://github.com/aws-amplify/amplify-android/blob/17fed06812c7a9ca65c6d47e7c5dbaed18846ed0/aws-api/src/main/java/com/amplifyframework/api/aws/auth/ApiRequestDecoratorFactory.java#L124-L130

There is an open issue in amplify-android repo seems relevant: https://github.com/aws-amplify/amplify-android/issues/1606
Will wait for amplify-android maintainer's investigation to proceed.

**Comment by two4zero** [comment]
now, reproduced errors.
not exactly... what i did...
1. in development. 'signed in' status. closed app in emulator or break debug in android studio.
2. reproduced errors.

so...
1. I deleted 'awsAPIPlugin.AdminQueries' field from 'amplifyconfiguration.dart'
2. and run debug
3. sign in and break debug
4. add 'awsAPIPlugin.AdminQueries' field
5. flutter clean.
6. run debug... and solved. working.

cc @Sunsiha Hope this helps you.

**Comment by aqarcoSys** [comment]
same here `CognitoUserSession is not valid because idToken is null` , I have an admin rest API also , any workaround ?
